src/rest-server/src/middlewares/v2/protocol.js

```diff
@@ -0,0 +1,117 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+
+// skip guard-for-in becasue protocolValidate has already filtered keys
+/* eslint guard-for-in: 0 */
+
+// module dependencies
+const yaml = require('js-yaml');
+const template = require('lodash.template');
+const protocolSchema = require('../../config/v2/protocol');
+
+const prerequisiteTypes = [
+    'script',
+    'output',
+    'data',
+    'dockerimage',
+];
+
+const prerequisiteFields = [
+    'script',
+    'output',
+    'data',
+    'dockerImage',
+];
+
+const protocolValidate = async (ctx, next) => {
+    // TODO
+    const protocolYAML = ctx.body;
+    const protocolJSON = yaml.safeLoad(protocolYAML);
+    if (!protocolSchema.validate(protocolJSON)) {
+        throw new Error(protocolSchema.validate.errors);
+    }
+    // convert prerequisites list to dict
+    const prerequisites = {};
+    for (let type of prerequisiteTypes) {
+        prerequisites[type] = {};
+    }
+    if ('prerequisites' in protocolJSON) {
+        for (let item of protocolJSON.prerequisites) {
+            prerequisites[item.type][item.name] = item;
+        }
+    }
+    protocolJSON.prerequisites = prerequisites;
+    // convert deployments list to dict
+    const deployments = {};
+    if ('deployments' in protocolJSON) {
+        for (let item of protocolJSON.deployments) {
+            deployments[item.name] = item;
+        }
+    }
+    protocolJSON.deployments = deployments;
+    // check prerequisites in taskRoles
+    for (let taskRole in protocolJSON.taskRoles) {
+        for (let field of prerequisiteFields) {
+            if (field in protocolJSON.taskRoles[taskRole] &&
+                !(protocolJSON.taskRoles[taskRole][field] in prerequisites[field.toLowerCase()])) {
+                throw new Error('prerequisite does not exist');
+            }
+        }
+    }
+    // check deployment in defaults
+    if ('defaults' in protocolJSON) {
+        if ('deployment' in protocolJSON.defaults &&
+            !(protocolJSON.defaults.deployment in deployments)) {
+            throw new Error('default deployment does not exist');
+        }
+    }
+    await next();
+};
+
+const protocolRender = async (ctx, next) => {
+    const protocolJSON = ctx.body;
+    for (let taskRole in protocolJSON.taskRoles) {
+        let commands = protocolJSON.taskRoles[taskRole].commands;
+        if (taskRole in protocolJSON.deployments) {
+            if ('preCommands' in protocolJSON.deployments[taskRole]) {
+                commands = protocolJSON.deployments[taskRole].preCommands.concat(commands);
+            }
+            if ('postCommands' in protocolJSON.deployments[taskRole]) {
+                commands = commands.concat(protocolJSON.deployments[taskRole].postCommands);
+            }
+        }
+        let entrypoint = commands.join(' ; ').replace('<%', '<%=');
+        const compiled = template(entrypoint);
```

Comment:
EJS-style template will introduce lots of vulnerabilities and could not be used for user's input.

For example: if user input `<% process.exit(1) %>`, it will kill the rest-server itself.

Reply:
Because the input string has been replaced `<%` with `<%=`, all strings inside "evaluate" delimiter won't be executed and will be treated as "interpolate" delimiter. In your case, nothing will happen.

Reply:
Change to `mustache`.